File: Java/swea/_1245_균형점.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class _1245_균형점 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());
        for (int tc = 1; tc <= T; tc++) {
            int N = Integer.parseInt(br.readLine());
            int[] pos = new int[N];
            int[] mass = new int[N];
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int i = 0; i < N * 2; i++) {
                if (i < N) pos[i] = Integer.parseInt(st.nextToken());
                else mass[i-N] = Integer.parseInt(st.nextToken());
            }
            double[] answer = new double[N-1];
            for (int i = 0; i < N-1; i++) {
                double left = pos[i];
                double right = pos[i+1];
                double mid = 0;
                while (right - left > 1 / Math.pow(10, 12)) {
                    mid = (left + right) / 2;
                    double leftPower = 0;
                    double rightPower = 0;
                    for (int j = 0; j < N; j++) {
                        if (pos[j] < mid) leftPower += mass[j] / Math.pow(mid-pos[j], 2);
                        else rightPower += mass[j] / Math.pow(pos[j]-mid, 2);
                    }
                    if (leftPower > rightPower) left = mid;
                    else right = mid;
                }
                answer[i] = mid;
            }
            System.out.print("#" + tc + " ");
            for (double balance : answer) {
                System.out.printf("%.10f ", balance);
            }
            System.out.println();
        }
    }
}

File: Python/swea/1245-균형점.py
T = int(input())
for tc in range(1, T + 1):
    n = int(input())
    arr = list(map(int, input().split()))
    pos = arr[:n]
    mass = arr[n:]
    answer = [0] * (n-1)
    for i in range(n-1):   # 균형점은 n-1개
        left = pos[i]
        right = pos[i + 1]
        while right - left > 1 / (10**12):  # 좌표 값의 오차가 10 ** -12 보다 작으면 빠져나감
            mid = (left + right) / 2
            left_power = 0
            right_power = 0
            for j in range(n):
                if pos[j] < mid:    # 왼쪽 자성체
                    left_power += mass[j] / (mid-pos[j])**2
                else:   # 오른쪽 자성체
                    right_power += mass[j] / (pos[j]-mid)**2
            if left_power > right_power:
                left = mid
            else:
                right = mid
        answer[i] = mid
    print(f'#{tc}', end=" ")
    for balance in answer:
        print(f'{balance:.10f}', end=" ")
    print()